src/analysis/normed_space/deriv.lean

```diff
@@ -0,0 +1,178 @@
+  has_fderiv_at_within f f' x s ↔
+    is_bounded_linear_map f' ∧
+      is_littleo (λ x', f x' - f x - f' (x' - x)) (λ x', x' - x) (nhds_within x s) :=
+and.congr_right_iff.mpr $
+  assume bf' : is_bounded_linear_map f',
+  have f'0 : f' 0 = 0 := bf'.to_linear_map.map_zero,
+  have h : ∀ x', ∥x' - x∥ = 0 → ∥f x' - f x - f' (x' - x)∥ = 0, from
+    assume x' hx',
+    have x' = x,
+      from eq_of_sub_eq_zero ((norm_eq_zero _).mp hx'),
+    by rw this; simp [f'0],
+  begin
+    rw has_fderiv_equiv_aux _ _ _ _ bf',
+    rw ←is_littleo_norm_left,
+    rw ←is_littleo_norm_right,
+    rw is_littleo_iff_tendsto h,
+    apply iff_of_eq, congr, ext x',
+    apply mul_comm
+  end
+
+def has_fderiv_at_within_mono {f : E → F} {f' : E → F} {x : E} {s t : set E}
+    (h : has_fderiv_at_within f f' x t) (hst : s ⊆ t) :
+  has_fderiv_at_within f f' x s :=
+and.intro h.left (tendsto_nhds_within_mono_left h.right hst)
+
+theorem has_fderiv_at_within.congr {f₀ f₁ : E → F} {f₀' f₁' : E → F}
+    {x : E} {s : set E} (h : has_fderiv_at_within f₀ f₀' x s)
+    (h₀ : ∀ x, f₀ x = f₁ x) (h₁ : ∀ x, f₀' x = f₁' x) :
+  has_fderiv_at_within f₁ f₁' x s :=
+by rw [←(funext h₀ : f₀ = f₁), ←(funext h₁ : f₀' = f₁')]; exact h
+
+theorem has_fderiv_id (x : E) (s : set E) : has_fderiv_at_within id id x s :=
+and.intro is_bounded_linear_map.id $
+  (@tendsto_const_nhds _ _ _ 0 _).congr (by simp)
+
+theorem has_fderiv_const (c : F) (x : E) (s : set E) :
+  has_fderiv_at_within (λ x, c) (λ y, 0) x s :=
+and.intro is_bounded_linear_map.zero $
+  (@tendsto_const_nhds _ _ _ 0 _).congr (by simp)
+
+theorem has_fderiv_smul {f : E → F} {f' : E → F} {x : E} {s : set E}
+    (c : ℝ) (h : has_fderiv_at_within f f' x s) :
+  has_fderiv_at_within (λ x, c • f x) (λ x, c • f' x) x s :=
+and.intro (is_bounded_linear_map.smul c h.left) $
+begin
+  have h' := tendsto_smul_const c h.right,
+  rw smul_zero at h',
+  apply h'.congr, intro x, simp,
+  rw [smul_smul, mul_comm c, ←smul_smul, smul_add, smul_add c, smul_neg, smul_neg]
+end
+
+theorem has_fderiv_add {f g : E → F} {f' g' : E → F} {x : E} {s : set E}
+    (hf : has_fderiv_at_within f f' x s) (hg : has_fderiv_at_within g g' x s) :
+  has_fderiv_at_within (λ x, f x + g x) (λ x, f' x + g' x) x s :=
+and.intro (is_bounded_linear_map.add hf.left hg.left) $
+begin
+  have h' := tendsto_add hf.right hg.right,
+  rw add_zero at h',
+  apply h'.congr, intro x, simp [smul_add]
+end
+
+theorem has_fderiv_neg {f : E → F} {f' : E → F} {x : E} {s : set E}
+    (h : has_fderiv_at_within f f' x s) :
+  has_fderiv_at_within (λ x, - f x) (λ x, - f' x) x s :=
+(has_fderiv_smul (-1 : ℝ) h).congr (by simp) (by simp)
+
+theorem has_fderiv_sub {f g : E → F} {f' g' : E → F} {x : E} {s : set E}
+    (hf : has_fderiv_at_within f f' x s) (hg : has_fderiv_at_within g g' x s) :
+  has_fderiv_at_within (λ x, f x - g x) (λ x, f' x - g' x) x s :=
+has_fderiv_add hf (has_fderiv_neg hg)
+
+theorem continuous_of_has_fderiv {f : E → F} {f' : E → F} {x : E} {s : set E}
```

Comment:
Might this proof be shortened by using o-analysis? This seems like a reasonable second touchstone for the o-analysis library other than the chain rule below.

Reply:
Done. Soon I'll update the PR with a nicer version.